miden/src/cli/debug.rs

```diff
@@ -0,0 +1,254 @@
+use super::data::{InputFile, ProgramFile};
+use crate::{
+    repl::{print_mem_address, print_stack},
+    VmState,
+};
+use miden::VmStateIterator;
+use rustyline::{Config, EditMode, Editor};
+use std::path::PathBuf;
+use structopt::StructOpt;
+
+#[derive(StructOpt, Debug)]
+#[structopt(name = "Debug", about = "Debug a miden program")]
+pub struct DebugCmd {
+    /// Path to .masm assembly file
+    #[structopt(short = "a", long = "assembly", parse(from_os_str))]
+    assembly_file: PathBuf,
+    /// Path to input file
+    #[structopt(short = "i", long = "input", parse(from_os_str))]
+    input_file: Option<PathBuf>,
+    /// Enable vi edit mode
+    #[structopt(short = "vi", long = "vim_edit_mode")]
+    vim_edit_mode: Option<String>,
+}
+
+impl DebugCmd {
+    pub fn execute(&self) -> Result<(), String> {
+        println!("============================================================");
+        println!("Debug program");
+        println!("============================================================");
+
+        // load program from file and compile
+        let program = ProgramFile::read(&self.assembly_file)?;
+
+        let program_hash: [u8; 32] = program.hash().into();
+        println!("Debugging program with hash {}... ", hex::encode(program_hash));
+
+        // load input data from file
+        let input_data = InputFile::read(&self.input_file, &self.assembly_file)?;
+
+        // fetch the stack and program inputs from the arguments
+        let stack_inputs = input_data.parse_stack_inputs()?;
+        let advice_provider = input_data.parse_advice_provider()?;
+
+        // generate debug trace
+        let mut vm_state_iter = processor::execute_iter(&program, stack_inputs, advice_provider);
+        let mut vm_state = vm_state_iter
+            .next()
+            .ok_or(format!("failed to start iterating vm state"))?
+            .unwrap();
+
+        // initializing readline
+        let mut rl_config = Config::builder().auto_add_history(true);
+
+        // add vim edit mode
+        if self.vim_edit_mode.is_some() {
+            rl_config = rl_config.edit_mode(EditMode::Vi);
+        }
+
+        // build config
+        let rl_config = rl_config.build();
+
+        let mut rl =
+            Editor::<()>::with_config(rl_config).expect("Readline couldn't be initialized");
+
+        loop {
+            match rl.readline(">> ") {
+                Ok(command) => match command.as_str() {
```

Comment:
The logic below is quite complex, in part because we are combining the logic of parsing and executing. I wonder if it would be better to separate the two. For example, we could parse a command into an enum in a stand-alone function and then here just match on that enum. The execution logic could be slightly simplified too because `next` could be represented as `play.1` etc.

The enum could look like this:
```Rust
enum DebugCommand {
    PlayAll,
    Play(usize),
    RewindAll,
    Rewind(usize),
    PrintState,
    PrintStack,
    PrintStackItem(usize),
    PrintMem,
    PrintMemAddress(usize),
    Clock,
    Quit,
    Help,
}

Reply:
I just had a small conversation with @Dominik1999 about the Playground, and this would also be very useful API to use in the debugger there. 

It would be nice to have that in a simple API that can be used via WASM

Reply:
yeah I'm building that now.  I should have a first draft built shortly.